### PR TITLE
Adding Kubernetes events support and events for various operations. #KO-48

### PR DESCRIPTION
**Pull request by tanmayja**

### CODE REVIEW

File: controllers/reconciler.go [hunk 1]
```diff
@@ -99,12 +114,16 @@ func (r *SingleClusterReconciler) Reconcile() (ctrl.Result, error) {
...
+		r.Recorder.Event(r.aeroCluster, corev1.EventTypeWarning, "ACLUpdateFailed",
+			fmt.Sprintf("Failed to setup Access Control %s/%s", r.aeroCluster.Namespace, r.aeroCluster.Name))
 		return reconcile.Result{}, err
 	}
 
 	// Update the AerospikeCluster status.
 	if err := r.updateStatus(); err != nil {
 		r.Log.Error(err, "Failed to update AerospikeCluster status")
+		r.Recorder.Event(r.aeroCluster, corev1.EventTypeWarning, "StatusUpdateFailed",
+			fmt.Sprintf("Failed to update AerospikeCluster status %s/%s", r.aeroCluster.Namespace, r.aeroCluster.Name))
```
Comment (sud82) [inline controllers/reconciler.go:126]: Check the above events. Use either cluster or AerospikeCluster

File: controllers/statefulset.go [hunk 2]
```diff
@@ -410,7 +411,15 @@ func (r *SingleClusterReconciler) buildSTSConfigMap(
 		return fmt.Errorf("failed to build config map data: %v", err)
 	}
 
-	// Replace config map data since we are supposed to create a new config map.
+	// Replace config map data if differs since we are supposed to create a new config map.
+	if reflect.DeepEqual(confMap.Data, configMapData) {
+		r.Log.Info(
+			"Already existed configmap is same as current configmap, so no change required",
```
Comment (sud82) [inline controllers/statefulset.go:417]: This log may become very frequent.
I think we can remove this log and add the log only if configMap is getting updated.

**Review by sud82 — CHANGES_REQUESTED** [review]

File: controllers/rack.go [hunk 3]
```diff
@@ -472,6 +502,9 @@ func (r *SingleClusterReconciler) scaleDownRack(
 		"ScaleDown AerospikeCluster statefulset", "desiredSz", desiredSize,
 		"currentSz", *found.Spec.Replicas,
 	)
+	r.Recorder.Event(r.aeroCluster, corev1.EventTypeNormal, "RackScaleDown",
```
Comment (sud82) [inline controllers/rack.go:505]: It's better to place this event and the above log when scale down is needed. Basically after this condition
`if *found.Spec.Replicas <= desiredSize {
		return found, reconcileSuccess()
	}`

File: controllers/reconciler.go [hunk 4]
```diff
@@ -51,7 +53,12 @@ func (r *SingleClusterReconciler) Reconcile() (ctrl.Result, error) {
 		r.Log.V(1).Info("Deleting AerospikeCluster")
 		// The cluster is being deleted
 		if err := r.handleClusterDeletion(finalizerName); err != nil {
+			r.Recorder.Event(r.aeroCluster, corev1.EventTypeWarning, "DeleteFailed",
+				fmt.Sprintf("Unable to handle AerospikeCluster delete operations %s/%s", r.aeroCluster.Namespace, r.aeroCluster.Name))
 			return reconcile.Result{}, err
+		} else {
+			r.Recorder.Event(r.aeroCluster, corev1.EventTypeNormal, "Deleted",
+				fmt.Sprintf("Deleted cluster %s/%s", r.aeroCluster.Namespace, r.aeroCluster.Name))
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply controllers/reconciler.go]: using AerospikeCluster now

**Review by sud82 — APPROVED** [review]

**Review by ashishshinde — APPROVED** [review]

**Pull request merged**